code/modules/integrated_electronics/core/analyzer.dm

```diff
@@ -0,0 +1,93 @@
+
+
+/obj/item/device/integrated_electronics/analyzer
+	name = "circuit analyzer"
+	desc = "This tool can scan an assembly and generate code necessary to recreate it in a circuit printer."
+	icon = 'icons/obj/electronic_assemblies.dmi'
+	icon_state = "analyzer"
+	flags_1 = CONDUCT_1
+	w_class = WEIGHT_CLASS_SMALL
+	var/list/circuit_list = list()
+	var/list/assembly_list = list(new /obj/item/device/electronic_assembly(null),
+			new /obj/item/device/electronic_assembly/medium(null),
+			new /obj/item/device/electronic_assembly/large(null),
+			new /obj/item/device/electronic_assembly/drone(null))
+
+/obj/item/device/integrated_electronics/analyzer/afterattack(var/atom/A, var/mob/living/user)
+	visible_message( "<span class='notice'>attempt to scan</span>")
+	if(ispath(A.type,/obj/item/device/electronic_assembly))
+		var/i = 0
+		var/j = 0
+		var/HTML ="start.assembly{{*}}"  //1-st in chapters.1-st block is just to secure start of program from excess symbols.{{*}} is delimeter for chapters.
+		visible_message( "<span class='notice'>start of scan</span>")
+		for(var/obj/item/I in assembly_list)
+			if( A.type == I.type )
+				HTML += I.name+"=-="+A.name         //2-nd block.assembly type and name. Maybe in future there will also be color and accesories.
+				break
+		/*
+		If(I.name == "electronic implant")
+			var/obj/item/weapon/implant/integrated_circuit/PI = PA        //now it can't recreate electronic implants.and devices maybe I'll fix it later.
+			var/obj/item/device/electronic_assembly/implant/PIC = PI.IC
+			A = PIC
+			*/
+		HTML += "{{*}}components"                   //3-rd block.components. First element is useless.delimeter for elements is ^%^.In element first circuit's default name.Second is user given name.delimiter is =-=
+
+		for(var/obj/item/integrated_circuit/IC in A.contents)
+			i =i + 1
+			HTML += "^%^"+IC.name+"=-="+IC.displayed_name
+		if(i == 0)
+			return
+		HTML += "{{*}}values"					//4-th block.values. First element is useless.delimeter for elements is ^%^.In element first i/o id.Second is data type.third is value.delimiter is :+:
+
+		i = 0
+		var/val
+		var/list/inp=list()
+		var/list/out=list()
+		var/list/act=list()
+		var/list/ioa=list()
+		for(var/obj/item/integrated_circuit/IC in A.contents)
+			i += 1
+			j = 0
+			for(var/datum/integrated_io/IN in IC.inputs)
```

Comment:
Do typeless loops when you know the contents are always going to be that type

Also the `in` keyword in loops like this duplicates the list before starting, so instead:

```
if(IC.inputs && IC.inputs.len) //take out the first bit if the list always exists
   for(var/k in 1 to IC.inputs.len)
      var/datum/integrated_io/IN = IC.inputs[k]
      inp[IN] = "[i]i[k]"
...
```

You also don't need the `j` iterator this way